File: containerWithMostWater.cpp
// two pointer
//time complexity -> O(n)
//space complexity -> O(1)
class Solution{
public:
    int containerWithMostWater(vector<int>& heights){
        int l = 0;
        int r = 0;
        int res = 0;

        while (l < r){
            int heightOfContainer = max(heights[l], heights[r]);
            int widthOfContainer = r - l;
            int waterInContainer = heightOfContainer * widthOfContainer;
            res = max(res, waterInContainer);

            if(l <= r){
                l++;
            }
            else{
                r++;
            }
        }
        return res;
    }
}

File: attempt3/containerWithMostWater.py
class Solution:
    def containerWithMostWater(self, heights: List[int]) -> int:
        l = 0
        r = len(heights) - 1
        res = 0

        while (l < r):
            width = r - l
            height = min(heights[l], heights[r])

            currentArea = width * height

            res = max(res, currentArea)

        return res
    
#time complexity -> O(n) while using a while loop
#space complexity -> O(1) becuase we only need the pointers and no data structure used

